languages/multi-language/src/main/java/de/jplag/multilang/MultiLanguageParser.java

```diff
@@ -29,8 +43,46 @@ public List<Token> parseFiles(Set<File> files, boolean normalize) throws Parsing
         return results;
     }
 
+    private void registerLanguageMaps(List<Language> languages) {
+        for (Language language : languages.stream().sorted(Comparator.comparing(it -> it.hasPriority() ? 1 : 0)).toList()) {
+            for (String suffix : language.suffixes()) {
+                if (language.hasPriority()) {
+                    if (this.languageMapPriority.containsKey(suffix)) {
+                        throw new IllegalStateException(
+                                String.format(ERROR_MULTIPLE_LANGUAGES, suffix, language.getName(), this.languageMapPriority.get(suffix).getName()));
+                    }
+                    this.languageMapPriority.put(suffix.toLowerCase(), language);
+                } else {
+                    if (this.languageMap.containsKey(suffix) && !this.languageMapPriority.containsKey(suffix)) {
+                        throw new IllegalStateException(
+                                String.format(ERROR_MULTIPLE_LANGUAGES, suffix, language.getName(), this.languageMap.get(suffix).getName()));
+                    }
+                    this.languageMap.put(suffix.toLowerCase(), language);
+                }
+            }
+
+        }
+    }
+
     private Optional<Language> findLanguageForFile(File file) {
-        return this.languages.stream().filter(language -> Arrays.stream(language.suffixes()).anyMatch(suffix -> file.getName().endsWith(suffix)))
-                .findFirst();
+        String name = file.getName();
+        String suffix = name.substring(name.lastIndexOf('.')).toLowerCase();
+
+        if (this.languageMapPriority.containsKey(suffix)) {
+            return Optional.of(this.languageMapPriority.get(suffix));
+        }
+
+        if (this.languageMap.containsKey(suffix)) {
+            return Optional.of(this.languageMap.get(suffix));
+        }
+
+        return Optional.empty();
+    }
+
+    private static void printWarning() {
+        if (!hasPrintedWarning) {
+            hasPrintedWarning = true;
+            LOG.warn(WARNING);
+        }
```

Comment:
Ideally, we would like to avoid keeping static state here. We could move the warning to the `MultiLanguage.parse()` call, which is only called once.